<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Cell Ranger RNA-seq data set — CellRanger • Chromium</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  

<meta property="og:title" content="Cell Ranger RNA-seq data set — CellRanger" />
<meta property="og:description" content="Contains UMI droplet-based single-cell RNA-seq data." />
<meta name="twitter:card" content="summary" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">Chromium</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/acidgenomics/Chromium">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Cell Ranger RNA-seq data set</h1>
    <small class="dont-index">Source: <a href='https://github.com/acidgenomics/Chromium/blob/master/R/AllGenerators.R'><code>R/AllGenerators.R</code></a></small>
    <div class="hidden name"><code>CellRanger.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Contains UMI droplet-based single-cell RNA-seq data.</p>
    </div>

    <pre class="usage"><span class='fu'>CellRanger</span>(
  <span class='no'>dir</span>,
  <span class='kw'>filtered</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>organism</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>ensemblRelease</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>genomeBuild</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>gffFile</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>refdataDir</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>samples</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>censorSamples</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>sampleMetadataFile</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>transgeneNames</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>interestingGroups</span> <span class='kw'>=</span> <span class='st'>"sampleName"</span>,
  <span class='kw'>BPPARAM</span> <span class='kw'>=</span> <span class='kw pkg'>BiocParallel</span><span class='kw ns'>::</span><span class='fu'><a href='https://rdrr.io/pkg/BiocParallel/man/register.html'>bpparam</a></span>()
)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>dir</th>
      <td><p><code><a href='https://rdrr.io/r/base/character.html'>character(1)</a></code>.
Directory path to Cell Ranger output.</p></td>
    </tr>
    <tr>
      <th>filtered</th>
      <td><p><code><a href='https://rdrr.io/r/base/logical.html'>logical(1)</a></code>.
Use filtered (recommended) or raw counts. Note that raw counts still
contain only whitelisted cellular barcodes.</p></td>
    </tr>
    <tr>
      <th>organism</th>
      <td><p><code><a href='https://rdrr.io/r/base/character.html'>character(1)</a></code>.
Full Latin organism name (e.g. <code>"Homo sapiens"</code>).</p></td>
    </tr>
    <tr>
      <th>ensemblRelease</th>
      <td><p><code><a href='https://rdrr.io/r/base/integer.html'>integer(1)</a></code>.
Ensembl release version (e.g. <code>90</code>). We recommend setting this value if
possible, for improved reproducibility. When left unset, the latest release
available via AnnotationHub/ensembldb is used. Note that the latest version
available can vary, depending on the versions of AnnotationHub and
ensembldb in use.</p></td>
    </tr>
    <tr>
      <th>genomeBuild</th>
      <td><p><code><a href='https://rdrr.io/r/base/character.html'>character(1)</a></code>.
Ensembl genome build assembly name (e.g. <code>"GRCh38"</code>). If set <code>NULL</code>,
defaults to the most recent build available.
Note: don't pass in UCSC build IDs (e.g. <code>"hg38"</code>).</p></td>
    </tr>
    <tr>
      <th>gffFile</th>
      <td><p><code><a href='https://rdrr.io/r/base/character.html'>character(1)</a></code>.
GFF/GTF (General Feature Format) file.
Generally, we recommend using GTF (GFFv2) instead of GFFv3.</p></td>
    </tr>
    <tr>
      <th>refdataDir</th>
      <td><p><code><a href='https://rdrr.io/r/base/character.html'>character(1)</a></code> or <code>NULL</code>.
Directory path to Cell Ranger reference annotation data.</p></td>
    </tr>
    <tr>
      <th>samples</th>
      <td><p><code>character</code>.
Sample identifiers.</p></td>
    </tr>
    <tr>
      <th>censorSamples</th>
      <td><p><code>character</code>.
Specify a subset of samples to censor.</p></td>
    </tr>
    <tr>
      <th>sampleMetadataFile</th>
      <td><p><code><a href='https://rdrr.io/r/base/character.html'>character(1)</a></code>.
Sample metadata file path. CSV or TSV is preferred, but Excel worksheets
are also supported. Check the documentation for conventions and required
columns.</p></td>
    </tr>
    <tr>
      <th>transgeneNames</th>
      <td><p><code>character</code>.
Vector indicating which assay rows denote transgenes (e.g. EGFP, TDTOMATO).</p></td>
    </tr>
    <tr>
      <th>interestingGroups</th>
      <td><p><code>character</code>.
Groups of interest to use for visualization.
Corresponds to factors describing the columns of the object.</p></td>
    </tr>
    <tr>
      <th>BPPARAM</th>
      <td><p><code>bpparamClass</code>.
BiocParallel parameter to specify the desired processor configuration.<br />
We recommend using one of the following:</p><ul>
<li><p><a href='https://rdrr.io/pkg/BiocParallel/man/register.html'>bpparam</a>.</p></li>
<li><p><a href='https://rdrr.io/pkg/BiocParallel/man/SerialParam-class.html'>SerialParam</a>.</p></li>
<li><p><a href='https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html'>MulticoreParam</a>.</p></li>
</ul></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p><code>CellRanger</code>.</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>Read <a href='https://www.10xgenomics.com/software/'>10x Genomics Cell Ranger</a> output
for a Chromium data set into a <code>SingleCellExperiment</code> object.</p>
    <h2 class="hasAnchor" id="note"><a class="anchor" href="#note"></a>Note</h2>

    <p>Currently supports loading of a single genome.</p>
<p>Updated 2020-02-24.</p>
    <h2 class="hasAnchor" id="directory-structure-for-multiple-samples"><a class="anchor" href="#directory-structure-for-multiple-samples"></a>Directory structure for multiple samples</h2>

    


<p>Cell Ranger can vary in its output directory structure, but we're requiring a
single, consistent directory structure for datasets containing multiple
samples that have not been aggregated into a single matrix with <code>aggr</code>.</p>
<p>Cell Ranger v3 output:</p><pre>| &lt;dir&gt;/
|-- &lt;sampleName&gt;/
|---- SC_RNA_COUNTER_CS/
|---- outs/
|------ filtered_feature_bc_matrix/
|-------- barcodes.tsv.gz
|-------- features.tsv.gz
|-------- matrix.mtx.gz
|------ filtered_feature_bc_matrix.h5
|------ metrics_summary.csv
|------ molecule_info.h5
|------ possorted_genome_bam.bam
|------ possorted_genome_bam.bam.bai
|------ raw_feature_bc_matrix/
|-------- barcodes.tsv.gz
|-------- features.tsv.gz
|-------- matrix.mtx.gz
|------ raw_feature_bc_matrix.h5
|------ web_summary.html
</pre>

<p>Cell Ranger v2 output:</p><pre>| &lt;dir&gt;/
|-- &lt;sampleName&gt;/
|---- SC_RNA_COUNTER_CS/
|---- outs/
|------ filtered_gene_bc_matrices/
|-------- &lt;genomeBuild&gt;/
|---------- barcodes.tsv
|---------- genes.tsv
|---------- matrix.mtx
|------ filtered_gene_bc_matrices_h5.h5
|------ metrics_summary.csv
|------ molecule_info.h5
|------ possorted_genome_bam.bam
|------ possorted_genome_bam.bam.bai
|------ raw_gene_bc_matrices/
|-------- &lt;genomeBuild&gt;/
|---------- barcodes.tsv
|---------- genes.tsv
|---------- matrix.mtx
|------ raw_gene_bc_matrices_h5.h5
</pre>

    <h2 class="hasAnchor" id="sample-metadata"><a class="anchor" href="#sample-metadata"></a>Sample metadata</h2>

    


<p>A user-supplied sample metadata file defined by <code>sampleMetadataFile</code> is
required for multiplexed datasets. Otherwise this can be left <code>NULL</code>, and
minimal sample data will be used, based on the directory names.</p>
    <h2 class="hasAnchor" id="reference-data"><a class="anchor" href="#reference-data"></a>Reference data</h2>

    


<p>We strongly recommend supplying the corresponding reference data required for
Cell Ranger with the <code>refdataDir</code> argument. It will convert the gene
annotations defined in the GTF file into a <code>GRanges</code> object, which get
slotted in <code><a href='https://rdrr.io/pkg/SummarizedExperiment/man/RangedSummarizedExperiment-class.html'>rowRanges()</a></code>. Otherwise, the
function will attempt to use the most current annotations available from
Ensembl, and some gene IDs may not match, due to deprecation in the current
Ensembl release.</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='no'>dir</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span>(<span class='st'>"extdata/cellranger_v3"</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>"Chromium"</span>)
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>CellRanger</span>(<span class='no'>dir</span>)</div><div class='output co'>#&gt; </div><div class='output co'>#&gt; <span class='message'><span style='color: #00BBBB;'>──</span><span> </span><span style='font-weight: bold;'>Chromium</span><span> </span><span style='color: #00BBBB;'>─────────────────────────────────────────────</span><span></span></div><div class='output co'>#&gt; <span class='message'>Importing Chromium single-cell RNA-seq run</span></div><div class='output co'>#&gt; </div><div class='output co'>#&gt; <span class='message'>── </span><span style='font-weight: bold;'>Run info</span><span> ──</span></div><div class='output co'>#&gt; </div><div class='output co'>#&gt; <span class='message'>1 sample detected:</span></div><div class='output co'>#&gt; <span class='message'>● pbmc</span></div><div class='output co'>#&gt; </div><div class='output co'>#&gt; <span class='message'>── </span><span style='font-weight: bold;'>Sample metadata</span><span> ──</span></div><div class='output co'>#&gt; </div><div class='output co'>#&gt; <span class='message'>── </span><span style='font-weight: bold;'>Counts</span><span> ──</span></div><div class='output co'>#&gt; </div><div class='output co'>#&gt; <span class='message'>→ Importing counts from 'matrix.mtx.gz' file.</span></div><div class='output co'>#&gt; <span class='message'>→ Importing </span><span style='color: #0000BB;'>matrix.mtx.gz</span><span> using </span><span style='color: #0000BB;'>Matrix</span><span>::</span><span style='color: #262626;background-color: #DADADA;'>`readMM()`</span><span>.</span></div><div class='output co'>#&gt; <span class='message'>→ Importing </span><span style='color: #0000BB;'>features.tsv.gz</span><span> using </span><span style='color: #0000BB;'>data.table</span><span>::</span><span style='color: #262626;background-color: #DADADA;'>`fread()`</span><span>.</span></div><div class='output co'>#&gt; <span class='message'>→ Importing </span><span style='color: #0000BB;'>barcodes.tsv.gz</span><span> using </span><span style='color: #0000BB;'>base</span><span>::</span><span style='color: #262626;background-color: #DADADA;'>`readLines()`</span><span>.</span></div><div class='output co'>#&gt; </div><div class='output co'>#&gt; <span class='message'>── </span><span style='font-weight: bold;'>Feature metadata</span><span> ──</span></div><div class='output co'>#&gt; </div><div class='output co'>#&gt; <span class='message'></span><span style='color: #BBBB00;'>!</span><span> Slotting empty ranges into </span><span style='color: #262626;background-color: #DADADA;'>`rowRanges()`</span><span>.</span></div><div class='output co'>#&gt; </div><div class='output co'>#&gt; <span class='message'>── </span><span style='font-weight: bold;'>Metrics</span><span> ──</span></div><div class='output co'>#&gt; </div><div class='output co'>#&gt; <span class='message'>→ Importing </span><span style='color: #0000BB;'>metrics_summary.csv</span><span> using </span><span style='color: #0000BB;'>data.table</span><span>::</span><span style='color: #262626;background-color: #DADADA;'>`fread()`</span><span>.</span></div><div class='output co'>#&gt; </div><div class='output co'>#&gt; <span class='message'>── </span><span style='font-weight: bold;'>Column data</span><span> ──</span></div><div class='output co'>#&gt; </div><div class='output co'>#&gt; <span class='message'>── </span><span style='font-weight: bold;'>Metadata</span><span> ──</span></div><div class='output co'>#&gt; </div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BBBB;'>ℹ</span><span> Filtered zero count rows and columns:</span>
#&gt; <span class='message'>- 46 / 100 rows (46%)</span>
#&gt; <span class='message'>- 92 / 100 columns (92%)</span></div><div class='output co'>#&gt; <span class='message'>→ Calculating 92 sample metrics.</span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #BBBB00;'>!</span><span> Calculating metrics without biotype information.</span>
#&gt; <span class='message'></span><span style='color: #262626;background-color: #DADADA;'>`rowRanges()`</span><span> required to calculate: </span><span style='color: #262626;background-color: #DADADA;'>`nCoding, nMito, mitoRatio`</span><span>.</span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BBBB;'>ℹ</span><span> Prefilter: 85 / 92 samples (92%).</span></div><div class='output co'>#&gt; </div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BB00;'>✓</span><span> Chromium single-cell RNA-seq run imported successfully.</span></div><div class='input'><span class='fu'><a href='https://rdrr.io/r/base/print.html'>print</a></span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; class: CellRanger 
#&gt; dim: 46 85 
#&gt; metadata(22): allSamples call ... version wd
#&gt; assays(1): counts
#&gt; rownames(46): ENSG00000008128 ENSG00000008130 ... ENSG00000272106
#&gt;   ENSG00000272512
#&gt; rowData names(0):
#&gt; colnames(85): AAACCCAAGAGACAAG AAACCCAAGGCCTAGA ... AACCTGAGTCGAGCTC
#&gt;   AACCTGAGTCTCTCAC
#&gt; colData names(8): sampleID sampleName ... log10FeaturesPerCount
#&gt;   mitoRatio
#&gt; reducedDimNames(0):
#&gt; spikeNames(0):
#&gt; altExpNames(0):</div></span></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      <li><a href="#value">Value</a></li>
      <li><a href="#details">Details</a></li>
      <li><a href="#note">Note</a></li>
      <li><a href="#directory-structure-for-multiple-samples">Directory structure for multiple samples</a></li>
      <li><a href="#sample-metadata">Sample metadata</a></li>
      <li><a href="#reference-data">Reference data</a></li>
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://steinbaugh.com/'>Michael Steinbaugh</a>, <a href='https://acidgenomics.com/'>Acid Genomics</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


